src/tokens/ERC1155.vy

```diff
@@ -111,65 +180,88 @@ def supportsInterface(interface_id: bytes4) -> bool:
-                          data: Bytes[1024]):
+def safeBatchTransferFrom(
+    owner: address,
+    to: address,
+    ids: DynArray[uint256, max_value(uint16)],
+    amounts: DynArray[uint256, max_value(uint16)],
+    data: Bytes[1024]
+):
     """
     @dev Batched version of `safeTransferFrom`.
     @notice Note that `ids` and `amounts` must have the
-            same length. Also, if `to` refers to a smart
-            contract, it must implement {IERC1155Receiver-onERC1155Received}
-            and return the acceptance magic value.
+        same length. Also, if `to` refers to a smart
+        contract, it must implement {IERC1155Receiver-onERC1155Received}
+        and return the acceptance magic value.
     @param owner The 20-byte address which previously
-           owned the token.
+        owned the token.
     @param to The 20-byte receiver address.
     @param ids The 32-byte array of token identifiers. Note
-           that the order and length must match the 32-byte
-           `amounts` array.
+        that the order and length must match the 32-byte
+        `amounts` array.
     @param amounts The 32-byte array of token amounts that are being
-           transferred. Note that the order and length must match
-           the 32-byte `ids` array.
+        transferred. Note that the order and length must match
+        the 32-byte `ids` array.
     @param data The maximum 1024-byte additional data
-           with no specified format.
+        with no specified format.
     """
-    pass
+    assert owner == msg.sender or self.isApprovedForAll[owner][msg.sender], "ERC1155: Not authorized to transfer"
+    assert to != empty(address), "ERC1155: transfer to the zero address"
+    assert len(ids) == len(amounts), "ERC1155: batch lengths mismatch"
+
+    for i in range(max_value(uint16)):
+        amount: uint256 = amounts[i]
+        id: uint256 = ids[i]
+        assert self._balances[owner][id] >= amount, "ERC1155: insufficient balance"
+        self._balances[owner][id] = unsafe_sub(self._balances[owner][id], amount)
+        self._balances[to][id] = unsafe_add(self._balances[to][id], amount)
+
+    log TransferBatch(msg.sender, owner, to, ids, amounts)
+    assert self._check_on_erc1155_batch_received(owner, to, ids, amounts, data), "ERC1155: transfer to non-ERC1155Receiver implementer"
 
 
 @external
 @view
 def balanceOf(owner: address, id: uint256) -> uint256:
     """
     @dev Returns the amount of tokens of token type
-         `id` owned by `owner`.
-    @notice Note that `owner` cannot be the zero
-            address.
+        `id` owned by `owner`.
     @param owner The 20-byte owner address.
     @param id The 32-byte identifier of the token.
     @return uint256 The 32-byte token amount owned
-            by `owner`.
+        by `owner`.
     """
-    return empty(uint256)
+    return self._balances[owner][id]
```

Comment:
We have to add a check here that the `owner` is not the zero address.

Reply:
I wasn't able to find anything in the [EIP1155 spec](https://eips.ethereum.org/EIPS/eip-1155) about reverting on `balanceOf(address(0))` like in ERC721. Since the zero address will always have a zero-balance for every token (burns don't transfer _to_ zero address, it only shows up as a "transfer" in event logs), would it be best to just implement the following instead?

```vy
def balanceOf(owner: address, id: uint256) -> uint256:
    if (owner == empty(address)): return empty(uint256)
    return self._balances[owner][id]
```

Reply:
You are right, but it's actually implied by the standard with the other assumptions IMO, so returning a valid value (even though it's 0) is not correct. But this is a discussion about conventions... 

I would prefer using `assert`, like that:
```vyper
def balanceOf(owner: address, id: uint256) -> uint256:
    assert owner != empty(address), "ERC1155: address zero is not a valid owner"
    return self._balances[owner][id]
```

Reply:
Makes sense. Will update